cdap-common/src/main/java/co/cask/cdap/common/resource/ResourceBalancerService.java

```diff
@@ -49,46 +47,44 @@
 import java.util.Collection;
 import java.util.Set;
 import java.util.concurrent.Executor;
-import javax.annotation.Nullable;
 
 /**
- * Metrics processor service that processes events from Kafka.
+ * A services that automatically balances resource assignments between its instances.
  */
-public final class KafkaMetricsProcessorService extends AbstractExecutionThreadService {
+public abstract class ResourceBalancerService extends AbstractExecutionThreadService {
 
-  private static final Logger LOG = LoggerFactory.getLogger(KafkaMetricsProcessorService.class);
+  private static final Logger LOG = LoggerFactory.getLogger(ResourceBalancerService.class);
 
-  private final CConfiguration conf;
+  private final String serviceName;
+  private final int partitionCount;
   private final LeaderElection election;
   private final ResourceCoordinatorClient resourceClient;
   private final DiscoveryService discoveryService;
-  private final KafkaMetricsProcessorServiceFactory metricsProcessorFactory;
+  private final ZKClient zk;
 
   private final SettableFuture<?> completion;
 
   private Cancellable cancelDiscoverable;
   private Cancellable cancelResourceHandler;
 
-  @Nullable
-  private MetricsContext metricsContext;
-
-  @Inject
-  public KafkaMetricsProcessorService(CConfiguration conf,
-                                      final ZKClientService zkClient,
-                                      DiscoveryService discoveryService,
-                                      final DiscoveryServiceClient discoveryServiceClient,
-                                      KafkaMetricsProcessorServiceFactory metricsProcessorFactory) {
-    this.conf = conf;
-    this.metricsProcessorFactory = metricsProcessorFactory;
+  protected ResourceBalancerService(String serviceName,
+                                 int partitionCount,
+                                 final ZKClientService zkClient,
+                                 DiscoveryService discoveryService,
+                                 final DiscoveryServiceClient discoveryServiceClient) {
+    this.serviceName = serviceName;
+    this.partitionCount = partitionCount;
     this.discoveryService = discoveryService;
 
+    this.zk = ZKClients.namespace(zkClient, "/" + serviceName);
```

Comment:
You might want to have a javadoc for this constructor, saying that the ZKClient passed in will get further namespaced by the service name. This is to prevent children classes do the unnecessary namespacing by themselves.